File: app/src/main/java/com/thecoolguy/rumaan/fileio/data/repository/Repository.java
package com.thecoolguy.rumaan.fileio.data.repository;

import android.app.Application;
import android.util.Log;
import com.thecoolguy.rumaan.fileio.data.db.UploadHistoryRoomDatabase;
import com.thecoolguy.rumaan.fileio.data.db.UploadItemDao;
import com.thecoolguy.rumaan.fileio.data.models.FileEntity;
import com.thecoolguy.rumaan.fileio.data.models.LocalFile;
import com.thecoolguy.rumaan.fileio.listeners.FileUploadListener;
import com.thecoolguy.rumaan.fileio.listeners.FileUploadProgressListener;
import org.jetbrains.annotations.NotNull;

/**
 * Handles the data part of the application
 * - network
 * - database
 */
public final class Repository implements FileUploadListener {

  private static final String TAG = "Repository";

  private static Repository sInstance;

  private final UploadItemDao mUploadItemDao;
  private final UploadHistoryRoomDatabase uploadHistoryRoomDatabase;

  private Repository(
      Application application) {
    uploadHistoryRoomDatabase = UploadHistoryRoomDatabase.getInstance(application);
    mUploadItemDao = uploadHistoryRoomDatabase.uploadItemDao();
  }

  public static Repository getInstance() {
    if (sInstance == null) {
      // How the hell did that happen?
      throw new NullPointerException();
    }
    return sInstance;
  }

  public static void setInstance(Application application) {
    if (sInstance == null) {
      sInstance = new Repository(application);
    }
  }

  public void upload(LocalFile localFile, FileUploadProgressListener progressListener) {
    Uploader.INSTANCE.uploadFile(localFile, this, progressListener);
  }

  private void save(FileEntity fileEntity) {
    // TODO: save to database
    onComplete(fileEntity);
  }


  @Override
  public void onFileUpload(@NotNull FileEntity fileEntity) {
    // save the file to database
    save(fileEntity);
  }

  @Override
  public void onFileUploadError(@NotNull Exception exception) {
    // TODO: log this error
  }

  @Override
  public void uploadProgress(int progress) {
    // Do nothing currently
  }

  @Override
  public void onComplete(@NotNull FileEntity fileEntity) {
    Log.i(TAG, "onComplete: " + fileEntity.toString());
  }
}

File: app/src/main/java/com/thecoolguy/rumaan/fileio/repository/Repository.kt
package com.thecoolguy.rumaan.fileio.repository

import android.app.Application
import com.thecoolguy.rumaan.fileio.data.db.DatabaseHelper
import com.thecoolguy.rumaan.fileio.data.db.UploadHistoryRoomDatabase
import com.thecoolguy.rumaan.fileio.data.db.UploadItemDao
import com.thecoolguy.rumaan.fileio.data.models.FileEntity
import com.thecoolguy.rumaan.fileio.data.models.LocalFile
import com.thecoolguy.rumaan.fileio.listeners.DatabaseCallback
import com.thecoolguy.rumaan.fileio.listeners.UploadListener
import com.thecoolguy.rumaan.fileio.network.Uploader
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.schedulers.Schedulers

object Repository : UploadListener, DatabaseCallback {
    override fun onComplete(fileEntity: FileEntity) {
    }

    override fun onSave(fileEntity: FileEntity, id: Long) {
        // on file entity saved to database callback
        observerList.forEach {
            it?.onComplete(fileEntity)
        }
    }

    // use this as context throughout
    lateinit var applicationContext: Application

    private val mDao: UploadItemDao by lazy {
        UploadHistoryRoomDatabase.getInstance(applicationContext)
                .uploadItemDao()
    }

    fun getDao() = mDao

    private val observerList = mutableListOf<UploadListener?>()

    fun addObserver(uploadListener: UploadListener) {
        if (!observerList.contains(uploadListener)) {
            observerList.add(uploadListener)
        }
    }

    private fun saveToDb(fileEntity: FileEntity) {
        val disposable = DatabaseHelper.saveToDatabase(fileEntity, mDao)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeBy(onSuccess = {
                    /* Notify all listeners */
                    observerList.forEach {
                        it?.onComplete(fileEntity)
                    }
                })
        DisposableBucket.add(disposable)
    }

    fun upload(localFile: LocalFile, uploadListener: UploadListener) {
        // add the listener to list
        addObserver(uploadListener)

        // this refers to DatabaseCallback
        Uploader.upload(localFile, this)
    }

    override fun onUpload(fileEntity: FileEntity) {
        saveToDb(fileEntity)
    }

    override fun progress(progress: Int) {
    }

}